RationaleMCP/0031/differences.md

```diff
@@ -1234,3 +1234,8 @@ Accordingly, a function component declaration which is neither input nor output
 
 The new annotation `protected = true` provides a standardized way to indicate that a component declaration in Flat Modelica comes from a protected section in the full Modelica model.
 See [`protected` annotation](annotations.md#protected).
+
+
+## Clock partitions
+
```

Comment:
```suggestion

Modelica requires that tool clock partition the equations, meaning that the clock for variables and equations can be inferred - in particular they are associated to a sub-clock that is part of a base-clock.
The sub- and super-sampling factors of the sub-clocks may be inferred (in a slightly complicated way), and in general the sub-clocks and base-clocks cuts across the instance hierarchy.

The idea is to remove the need for that in Flat Modelica simplifying the analysis.
Note that if we want to extend Flat Modelica to be used as sub-components this implies that we have to decide whether to clock the component or not; that is similar to the need for external sampling in eFMI.

### Separate declaration of clock partitions

Introduce a new syntax for stating that equations and variables belong to a sub-clock. The variables are repeated, but the equations are introduced in this form.

A rough possibility would be (ignore the syntax - main idea is what to list):
Modelica:
model M
   Real x, y;
equation
   when Clock(1e-3) then
      x=sample(time);
    end when;
    y=subSample(x, 2);
```
Converted to something like this:
```
  BaseClock: Clock(1e-3);
  SubClock: subSample(BaseClock, 2)
     equation
     'y'='x';
     algorithm
      ...
      variables 'y';
   SubClock: BaseClock
     equation
       'x'='time';
     algorithm
      ...
      variables 'x';
```
Note that we can safely remove any "when Clock... then" from the equations.

#### Disadvantages:

Since clock-partitioning operators can take expressions as argument we ideally should declare the clock of sub-expressions (especially for clock-partitioning operators as argument to clock-partitioning operators). Tools can circumvent that by introducing additional variables.

Another option would be to "annotate" expressions with their clock - it avoids that problem, but would look messier.
```

Reply:
Is it really necessary to list the variables, belonging to a sub clock? Instead of
`y=x;`
and
`x=time;`
one could write
`y=noClock(x);`
or
`x=sample(time);`
or similar. And then it would be clear that the variables come from another partition.
Only variables which are referenced outside such special operators belong to the current clock.

This would only require a tool to run through the equations of a partition once and collect the belonging variables.

Reply:
Assuming that works in all cases we then have the possibility to only give a sub-clock for variables _or_ equations, and I can understand that listing equations under sub-clocks avoid repetitions.
On the other hand redundancy is sometimes good.